core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,296 @@
+        if (field.typeSignature().type() == TypeSignatureType.STRUCT && visited.containsKey(
+                fieldTypeName)) {
+            // If field is already visited, add a reference to the field instead of iterating its children.
+            final String pathName = visited.get(fieldTypeName);
+            fieldNode.put("$ref", pathName);
+        } else {
+            // Field is not visited, create a new type definition for it.
+            fieldNode.put("type", getSchemaType(field.typeSignature()));
+            fieldNode.put("description", field.descriptionInfo().docString());
+
+            // TODO: Allow maps to have custom types for keys and values.
+            // Currently they are allowed to take any additional properties.
+            fieldNode.put("additionalProperties", field.typeSignature().type() == TypeSignatureType.MAP);
+
+            // Fill required fields for the current object.
+            if (field.requirement() == FieldRequirement.REQUIRED) {
+                required.add(field.name());
+            }
+
+            if (field.typeSignature().type() == TypeSignatureType.ENUM) {
+                fieldNode.set("enum", getEnumType(field.typeSignature()));
+            }
+
+            if (field.typeSignature().type() == TypeSignatureType.ITERABLE) {
+                final ObjectNode itemsType = getItemsType(field.typeSignature(), path);
+                if (itemsType != null) {
+                    fieldNode.set("items", itemsType);
+                }
+            }
+
+            final StructInfo fieldStructInfo = typeNameToStructMapping.get(field.typeSignature().name());
+
+            // Set the current path to be "PREVIOUS_PATH/field.name"
+            final String currentPath = path + '/' + field.name();
+            // Mark current field as visited
+            visited.put(fieldTypeName, currentPath);
+
+            // Iterate over each child field, generate their definitions.
+            if (fieldStructInfo != null && !fieldStructInfo.fields().isEmpty()) {
+                generateFields(fieldStructInfo.fields(), visited, currentPath, fieldNode);
+            }
+        }
+
+        // Set current field inside the returned object.
+        root.set(field.name(), fieldNode);
+    }
+
+    /**
+     * Generate properties for the given fields and writes to the object node.
+     * @param fields list of fields that the child has.
+     * @param visited a map of visited fields, required for cycle detection.
+     * @param path current path as defined in JSON Schema spec, required for cyclic references.
+     * @param root object node that the results will be written to.
+     */
+    private void generateFields(List<FieldInfo> fields, Map<String, String> visited, String path,
+                                ObjectNode root) {
+        final ObjectNode objectNode = mapper.createObjectNode();
+        final ArrayNode required = mapper.createArrayNode();
+
+        // TODO: Consider filtering header & path params.
```

Comment:
Shouldn't be an issue for grpc rollout again.

Reply:
I think we can contain the fields whose `FieldLocation`s are `BODY` and `UNSPECIFIED`.